File: ml/hokm/game.py
import random
from enum import Enum, auto
from typing import List, Optional, Tuple, Dict

class Suit(Enum):
    HEARTS = auto()
    DIAMONDS = auto()
    CLUBS = auto()
    SPADES = auto()

    def __str__(self):
        return self.name

class Rank(Enum):
    TWO = 2
    THREE = 3
    FOUR = 4
    FIVE = 5
    SIX = 6
    SEVEN = 7
    EIGHT = 8
    NINE = 9
    TEN = 10
    JACK = 11
    QUEEN = 12
    KING = 13
    ACE = 14

    def __str__(self):
        return self.name

class Card:
    def __init__(self, rank: Rank, suit: Suit):
        self.rank = rank
        self.suit = suit

    def __repr__(self):
        return f"{self.rank.name} of {self.suit.name}"

    def __eq__(self, other):
        return self.rank == other.rank and self.suit == other.suit

    def __hash__(self):
        return hash((self.rank, self.suit))

class Player:
    def __init__(self, player_id: int):
        self.player_id = player_id
        self.hand: List[Card] = []
        self.won_tricks: int = 0

    def add_cards(self, cards: List[Card]):
        self.hand.extend(cards)
        self.hand.sort(key=lambda c: (c.suit.value, c.rank.value))

    def remove_card(self, card: Card):
        self.hand.remove(card)

    def has_suit(self, suit: Suit) -> bool:
        return any(c.suit == suit for c in self.hand)

class GameState(Enum):
    DETERMINE_HAKIM = auto()
    DECLARE_TRUMP = auto()
    PLAYING = auto()
    FINISHED = auto()

class HokmGame:
    def __init__(self):
        self.players = [Player(i) for i in range(4)]
        self.deck = self._create_deck()
        self.hakim_index = 0 # Default, will be determined
        self.trump_suit: Optional[Suit] = None
        self.current_player_index = 0
        self.state = GameState.DETERMINE_HAKIM
        self.current_trick: List[Tuple[int, Card]] = [] # (player_id, card)
        self.tricks_history: List[List[Tuple[int, Card]]] = []
        self.scores = {0: 0, 1: 0} # Team 0 (P0, P2) vs Team 1 (P1, P3)
        self.game_winner = None

    def _create_deck(self) -> List[Card]:
        deck = [Card(rank, suit) for suit in Suit for rank in Rank]
        random.shuffle(deck)
        return deck

    def reset(self):
        self.deck = self._create_deck()
        for p in self.players:
            p.hand = []
            p.won_tricks = 0
        self.trump_suit = None
        self.current_trick = []
        self.tricks_history = []
        self.state = GameState.DETERMINE_HAKIM
        # Hakim stays same unless specified otherwise, but for now let's redetermine or rotate?
        # Standard Hokm: Loser deals, Winner is Hakim.
        # For simplicity in ML env, we might randomize or keep fixed.
        # Let's implement a simple random deal for now.
        self._deal_initial_cards()

    def _deal_initial_cards(self):
        # Deal 5 cards to each
        for i, player in enumerate(self.players):
            player.add_cards(self.deck[i*5 : (i+1)*5])
        
        # If we need to determine Hakim based on first ace/highest card logic:
        # For this implementation, let's assume Hakim is already set or P0 for simplicity
        # unless we want to implement the full "deal until Ace" logic.
        # Let's stick to P0 is Hakim for the first game, then winner becomes Hakim.
        self.state = GameState.DECLARE_TRUMP
        self.current_player_index = self.hakim_index

    def declare_trump(self, suit: Suit):
        if self.state != GameState.DECLARE_TRUMP:
            raise ValueError("Not in trump declaration phase")
        if self.current_player_index != self.hakim_index:
            raise ValueError("Only Hakim can declare trump")
        
        self.trump_suit = suit
        self._deal_remaining_cards()
        self.state = GameState.PLAYING
        self.current_player_index = self.hakim_index # Hakim leads first trick

    def _deal_remaining_cards(self):
        # Deal remaining 8 cards
        start = 20
        for i, player in enumerate(self.players):
            player.add_cards(self.deck[start + i*8 : start + (i+1)*8])

    def get_legal_moves(self, player_id: int) -> List[Card]:
        player = self.players[player_id]
        if not self.current_trick:
            return player.hand
        
        lead_suit = self.current_trick[0][1].suit
        if player.has_suit(lead_suit):
            return [c for c in player.hand if c.suit == lead_suit]
        
        return player.hand

    def play_card(self, player_id: int, card: Card):
        if self.state != GameState.PLAYING:
            raise ValueError("Not in playing phase")
        if player_id != self.current_player_index:
            raise ValueError(f"Not player {player_id}'s turn")
        
        legal_moves = self.get_legal_moves(player_id)
        if card not in legal_moves:
             # This check assumes card object identity or equality. 
             # Since we reconstruct cards, equality check in Card class is important.
             if card not in player.hand:
                 raise ValueError("Card not in hand")
             raise ValueError(f"Illegal move: must follow suit {self.current_trick[0][1].suit}")

        self.players[player_id].remove_card(card)
        self.current_trick.append((player_id, card))
        
        if len(self.current_trick) == 4:
            self._resolve_trick()
        else:
            self.current_player_index = (self.current_player_index + 1) % 4

    def _resolve_trick(self):
        winner_idx = self._determine_trick_winner()
        self.players[winner_idx].won_tricks += 1
        
        # Team scoring
        team_idx = winner_idx % 2
        # We don't update game score yet, just track tricks. 
        # Game score is updated at end of round (13 tricks).
        
        self.tricks_history.append(self.current_trick)
        self.current_trick = []
        self.current_player_index = winner_idx
        
        if len(self.players[0].hand) == 0:
            self._end_round()

    def _determine_trick_winner(self) -> int:
        lead_suit = self.current_trick[0][1].suit
        best_play = self.current_trick[0]
        
        for player_id, card in self.current_trick[1:]:
            # If current best is trump
            if best_play[1].suit == self.trump_suit:
                if card.suit == self.trump_suit and card.rank.value > best_play[1].rank.value:
                    best_play = (player_id, card)
            # If current best is not trump
            else:
                if card.suit == self.trump_suit:
                    best_play = (player_id, card)
                elif card.suit == lead_suit and card.rank.value > best_play[1].rank.value:
                    best_play = (player_id, card)
                    
        return best_play[0]

    def _end_round(self):
        team0_tricks = self.players[0].won_tricks + self.players[2].won_tricks
        team1_tricks = self.players[1].won_tricks + self.players[3].won_tricks
        
        if team0_tricks >= 7:
            self.scores[0] += 1
            self.hakim_index = 0 # Or keep same team? Standard rules: winner becomes Hakim.
            # If team 0 was not Hakim, they become Hakim.
        else:
            self.scores[1] += 1
            self.hakim_index = 1
            
        # Check for match win (usually 7 points)
        # For ML training, we might just play single games or shorter matches.
        self.state = GameState.FINISHED

File: backend/src/game.rs
use serde::{Deserialize, Serialize};
use std::fmt;
use rand::seq::SliceRandom;
use rand::thread_rng;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum Suit {
    Hearts,
    Diamonds,
    Clubs,
    Spades,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, PartialOrd, Ord)]
pub enum Rank {
    Two = 2,
    Three = 3,
    Four = 4,
    Five = 5,
    Six = 6,
    Seven = 7,
    Eight = 8,
    Nine = 9,
    Ten = 10,
    Jack = 11,
    Queen = 12,
    King = 13,
    Ace = 14,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct Card {
    pub rank: Rank,
    pub suit: Suit,
}

impl fmt::Display for Card {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?} of {:?}", self.rank, self.suit)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Player {
    pub id: usize,
    pub hand: Vec<Card>,
    pub won_tricks: u32,
}

impl Player {
    pub fn new(id: usize) -> Self {
        Self {
            id,
            hand: Vec::new(),
            won_tricks: 0,
        }
    }

    pub fn add_cards(&mut self, cards: Vec<Card>) {
        self.hand.extend(cards);
        // Sort logic could be added here
    }

    pub fn remove_card(&mut self, card: &Card) -> bool {
        if let Some(pos) = self.hand.iter().position(|c| c == card) {
            self.hand.remove(pos);
            true
        } else {
            false
        }
    }

    pub fn has_suit(&self, suit: Suit) -> bool {
        self.hand.iter().any(|c| c.suit == suit)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum GameState {
    DetermineHakim,
    DeclareTrump,
    Playing,
    Finished,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HokmGame {
    pub players: Vec<Player>,
    pub hakim_index: usize,
    pub trump_suit: Option<Suit>,
    pub current_player_index: usize,
    pub state: GameState,
    pub current_trick: Vec<(usize, Card)>,
    pub scores: (u32, u32), // Team 0 (0,2) vs Team 1 (1,3)
}

impl HokmGame {
    pub fn new() -> Self {
        Self {
            players: (0..4).map(Player::new).collect(),
            hakim_index: 0, // Default
            trump_suit: None,
            current_player_index: 0,
            state: GameState::DetermineHakim,
            current_trick: Vec::new(),
            scores: (0, 0),
        }
    }

    pub fn create_deck() -> Vec<Card> {
        let mut deck = Vec::new();
        for suit in [Suit::Hearts, Suit::Diamonds, Suit::Clubs, Suit::Spades] {
            for rank in [
                Rank::Two, Rank::Three, Rank::Four, Rank::Five, Rank::Six, Rank::Seven,
                Rank::Eight, Rank::Nine, Rank::Ten, Rank::Jack, Rank::Queen, Rank::King, Rank::Ace,
            ] {
                deck.push(Card { rank, suit });
            }
        }
        let mut rng = thread_rng();
        deck.shuffle(&mut rng);
        deck
    }

    pub fn start_game(&mut self) {
        let deck = Self::create_deck();
        // Deal 5 cards
        for i in 0..4 {
            self.players[i].add_cards(deck[i*5..(i+1)*5].to_vec());
        }
        // Assume Hakim is 0 for now or logic to determine
        self.state = GameState::DeclareTrump;
        self.current_player_index = self.hakim_index;
        
        // Store remaining deck for later? In a real implementation we need to store it.
        // For this MVP let's just re-shuffle or store it in the struct (omitted for brevity but needed).
        // Let's assume we just deal all for now to simplify.
        let remaining = deck[20..].to_vec();
        // Hack: store remaining in player 0's extra field or just deal them now but hide them?
        // Correct way: Add `deck` field to struct.
        // For now, let's just deal all cards immediately but enforce phases.
         for i in 0..4 {
            self.players[i].add_cards(remaining[i*8..(i+1)*8].to_vec());
        }
    }

    pub fn declare_trump(&mut self, suit: Suit) -> Result<(), String> {
        if self.state != GameState::DeclareTrump {
            return Err("Not in trump declaration phase".to_string());
        }
        self.trump_suit = Some(suit);
        self.state = GameState::Playing;
        self.current_player_index = self.hakim_index;
        Ok(())
    }

    pub fn play_card(&mut self, player_id: usize, card: Card) -> Result<(), String> {
        if self.state != GameState::Playing {
            return Err("Not in playing phase".to_string());
        }
        if player_id != self.current_player_index {
            return Err(format!("Not player {}'s turn", player_id));
        }

        let player = &self.players[player_id];
        if !player.hand.contains(&card) {
            return Err("Card not in hand".to_string());
        }

        // Validate suit following
        if let Some((_, lead_card)) = self.current_trick.first() {
            if card.suit != lead_card.suit && player.has_suit(lead_card.suit) {
                return Err(format!("Must follow suit {:?}", lead_card.suit));
            }
        }

        self.players[player_id].remove_card(&card);
        self.current_trick.push((player_id, card));

        if self.current_trick.len() == 4 {
            self.resolve_trick();
        } else {
            self.current_player_index = (self.current_player_index + 1) % 4;
        }

        Ok(())
    }

    fn resolve_trick(&mut self) {
        let winner_idx = self.determine_trick_winner();
        self.players[winner_idx].won_tricks += 1;
        
        self.current_trick.clear();
        self.current_player_index = winner_idx;
        
        // Check end of round
        if self.players[0].hand.is_empty() {
            self.end_round();
        }
    }

    fn determine_trick_winner(&self) -> usize {
        let lead_suit = self.current_trick[0].1.suit;
        let mut best_play = self.current_trick[0];

        for &(player_id, card) in self.current_trick.iter().skip(1) {
            if let Some(trump) = self.trump_suit {
                if best_play.1.suit == trump {
                    if card.suit == trump && card.rank > best_play.1.rank {
                        best_play = (player_id, card);
                    }
                } else {
                    if card.suit == trump {
                        best_play = (player_id, card);
                    } else if card.suit == lead_suit && card.rank > best_play.1.rank {
                        best_play = (player_id, card);
                    }
                }
            } else {
                // No trump (shouldn't happen in Hokm but safe fallback)
                if card.suit == lead_suit && card.rank > best_play.1.rank {
                    best_play = (player_id, card);
                }
            }
        }
        best_play.0
    }

    fn end_round(&mut self) {
        let team0_tricks = self.players[0].won_tricks + self.players[2].won_tricks;
        if team0_tricks >= 7 {
            self.scores.0 += 1;
        } else {
            self.scores.1 += 1;
        }
        self.state = GameState::Finished;
    }
}
